jvm/middleware/src/main/kotlin/io/spine/logging/jvm/MutableMetadata.kt

```diff
@@ -0,0 +1,163 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.logging.jvm
+
+import io.spine.logging.jvm.backend.Metadata
+
+/**
+ * A [Metadata] with read/write features.
+ */
+internal class MutableMetadata : Metadata() {
+
+    private companion object {
+
+        /**
+         * The default number of key/value pairs we initially allocate space for when
+         * someone adds metadata to this context.
+         *
+         * Note: As of 10/12 the VM allocates small object arrays very linearly with
+         * respect to the number of elements.
+         * An array has a 12-byte header with 4 bytes/element for object references.
+         * The allocation size is always rounded up to the next 8 bytes, which means we
+         * can just pick a small value for the initial size and grow from there without
+         * too much waste.
+         *
+         * For 4 key/value pairs, we will need 8 elements in the array, which will
+         * take up 48 bytes `(12 + (8 * 4) = 44`, which when rounded up is 48.
+         */
+        private const val INITIAL_KEY_VALUE_CAPACITY = 4
+    }
+
+    /**
+     * The array of key/value pairs to hold any metadata that might
+     * be added by the logger or any of the fluent methods on our API.
+     *
+     * This is an array, so it is as space-efficient as possible.
+     */
+    private var keyValuePairs: Array<Any?> = arrayOfNulls(2 * INITIAL_KEY_VALUE_CAPACITY)
+
+    /**
+     * The number of key/value pairs currently stored in the array.
+     */
+    private var keyValueCount = 0
+
+    override fun size(): Int = keyValueCount
+
+    @Suppress("UNCHECKED_CAST")
+    override fun getKey(n: Int): MetadataKey<Any> {
+        if (n >= keyValueCount) {
+            throw IndexOutOfBoundsException(n)
+        }
+        return keyValuePairs[2 * n] as MetadataKey<Any>
+    }
+
+    override fun getValue(n: Int): Any {
+        if (n >= keyValueCount) {
+            throw IndexOutOfBoundsException(n)
+        }
+        return keyValuePairs[(2 * n) + 1]!!
+    }
+
+    private fun indexOf(key: MetadataKey<*>): Int {
+        for (index in 0 until keyValueCount) {
+            if (keyValuePairs[2 * index] == key) {
+                return index
+            }
+        }
+        return -1
+    }
+
+    override fun <T : Any> findValue(key: MetadataKey<T>): T? {
+        val index = indexOf(key)
+        return if (index != -1) key.cast(keyValuePairs[(2 * index) + 1]) else null
+    }
+
+    /**
+     * Adds the key/value pair to the metadata (growing the internal array as necessary). If the key
```

Comment:
A suggestion for the agentic intrusion for the future: 

Review the documentation in KDoc, make the first paragraph to be exactly one sentence, moving the rest of the text to the new paragraph.

Reply:
Added to the guidelines.